### PR TITLE
Write the template to a buffer to catch render errors

### LINKED ISSUE #3559: superfluous response.WriteHeader call in httplib
#### What is the problem you are having with rclone?
rclone serve http throws errors when using _lots_ of connections

#### What is your rclone version (output from `rclone version`)
1.49.0

#### Which OS you are using and how many bits (eg Windows 7, 64 bit)
Debian 64bit

####  Which cloud storage system are you using? (eg Google Drive)
Local

#### The command you were trying to run (eg `rclone copy /tmp remote:tmp`)
rclone serve http local: --transfers=9999 --checkers=9999

#### A log from the command with the `-vv` flag (eg output from `rclone -vv copy /tmp remote:tmp`)
http: superfluous response.WriteHeader call from github.com/rclone/rclone/cmd/serve/httplib/serve.Error (dir.go:72)

### PR DESCRIPTION
**Pull request by fionera**
#### What is the purpose of this change?
Fix #3559

#### Was the change discussed in an issue or in the forum before?
#3559

#### Checklist

- [x] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [x] I have added tests for all changes in this PR if appropriate.
- [x] I have added documentation for the changes if appropriate.
- [x] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [x] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

File: cmd/serve/httplib/serve/dir.go [hunk 1]
```diff
@@ -80,9 +83,14 @@ func (d *Directory) Serve(w http.ResponseWriter, r *http.Request) {
...
-	err := d.HTMLTemplate.Execute(w, d)
+	buf := &bytes.Buffer{}
+	err := d.HTMLTemplate.Execute(buf, d)
 	if err != nil {
 		Error(d.DirRemote, w, "Failed to render template", err)
 		return
 	}
+	_, err = buf.WriteTo(w)
+	if err != nil {
+		Error(d.DirRemote, nil, "Failed to drain template buffer", err)
```
Comment (terorie) [inline cmd/serve/httplib/serve/dir.go]: This doesn't solve the original problem. As soon as the first `w.Write` attempt has been made, `w.WriteHeader` was automatically called so `serve.Error` shouldn't be called in this case.

Comment (fionera) [reply cmd/serve/httplib/serve/dir.go]: serve.Error wont write to the http response tho, instead its only in the log

Comment (terorie) [reply cmd/serve/httplib/serve/dir.go]: Maybe update the doc then? :) `// Error returns an http.StatusInternalServerError and logs the error`

Comment (fionera) [reply cmd/serve/httplib/serve/dir.go]: Done :smile:

**Comment by terorie** [comment]
Quick idea – We can also do zero allocs if we create a custom `io.Writer` that remembers if any write occurred.

```go
type WriteChecker struct {
    written bool
}

func (w *WriteChecker) Write(p []byte) (n int, err error) {
    w.written = true
    return len(p), nil
}

func (w *WriteChecker) HasWritten() bool {
    return w.written
}

// ...

var checker WriteChecker
err := d.HTMLTemplate.Execute(buf, io.MultiWriter(&checker, w))
if err != nil {
    // ...
    !if checker.HasWritten() {
        serve.Error(...)
    }
}
```

**Comment by ncw** [comment]
It seems wrong to turn a streaming process into something which uses a memory buffer just to avoid an error message.

However I can't think of anything better so I'll merge this  - thank you :-)

**Pull request merged**